### PR TITLE
Switch painless dynamic calls to invokedynamic, remove perf hack/cheat

### PR DESCRIPTION
**Pull request by rmuir**
Currently using any dynamic language features of painless is slow. We even have in docs recommending people add a lot of casts to work around this. The hack in https://github.com/elastic/elasticsearch/pull/18169 speeds up access to a document's fields, but this is fragile and not a general solution.

This PR switches all dynamic calls, dynamic loads, and dynamic stores over to use invokedynamic with inline caching (cascaded, goes megamorphic at 5 types). I based it on this one (BSD license) https://github.com/qmx/jsr292-cookbook/blob/master/inlining-cache/src/jsr292/cookbook/icache/RT.java

We don't need to do anything with dynamic array load/store, its already a very simple 3-way decision tree and never involves a dynamic lookup. Same goes for the operators: i left that stuff as is.

I removed the instanceof-performance hack for accessing a document's fields, we don't need it anymore.

I cleaned up some getter/setter stuff too, we compute these at compile time. Also its now possible to get access to `isEmpty()`-style methods as `.empty`, we need that to support the scripting api the way its documented for groovy at least. 

For the nightly benchmark script I see:
SEARCH expression (median): 0.271813 sec
SEARCH painless_static (median): 0.517333 sec
SEARCH painless_dynamic (median): 0.582639 sec

So we should clean up and simplify the documentation, I don't think the dynamic access is trappy anymore, and we shouldn't recommend people add a bunch of casts to their scripts to try to make them faster. If someone wants to eek out extra speed, instead they should use lucene expressions which bypasses the slow scripting API.

### CODE REVIEW

**Comment by rmuir** [comment]
I cleaned up and ran benchmarks: this is currently a script that runs on all of geonames and looks like this: `(Math.log(Math.abs(input.doc['population'].value)) + input.doc['elevation'].value * input.doc['latitude'].value)/_score`

I altered the way we access document fields and values to show perf for different dynamic language features:

| style | median time (master) | median time (branch) |
| --- | --- | --- |
|`doc['field'].value`|0.631238|0.582639|
|`doc['field'].getValue()`|4.148362|0.609807|
|`doc.field.0`|5.567170|0.603587|
|`doc.field.get(0)`|6.588988|0.606248|

`.value` shows the hack from #18169 working: we remove the hack here and have the same performance. 
`.getValue()`: no-arg dynamic method
`.field.0`: two field loads, one that was especially slow before (the .0)
`.field.get(0)`: field load + dynamic method with parameter

The static version (with casts) is ~ `0.517333 s`, so there isn't much overhead. I benchmark with the scripts `inline` so these times include compilation and JIT costs.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/DynamicCallSite.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                return Def.lookupSetter(clazz, name, Definition.INSTANCE);
+            default: throw new AssertionError();
+        }
+    }
+    
+    /**
+     * Called when a new type is encountered (or, when we have encountered more than {@code MAX_DEPTH}
+     * types at this call site and given up on caching). 
+     */
+    public static Object fallback(InliningCacheCallSite callSite, Object[] args) throws Throwable {
```
Comment (uschindler) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/DynamicCallSite.java]: This method can be private, as we only refer via methodhandle.
If compiler emits warning, add a @SuppressWarnings to it

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/DynamicCallSite.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        
+        callSite.setTarget(fallback);
+        return callSite;
+    }
+    
+    /** 
+     * guard method for inline caching: checks the receiver's class is the same
+     * as the cached class
+     */
+    public static boolean checkClass(Class<?> clazz, Object receiver) {
```
Comment (uschindler) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/DynamicCallSite.java]: can be private, too (see below)

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java [hunk 3]
```diff
@@ -19,371 +19,302 @@
...
+    /** pointer to Map.get(Object) */
+    private static final MethodHandle MAP_GET;
+    /** pointer to Map.put(Object,Object) */
+    private static final MethodHandle MAP_PUT;
+    /** pointer to List.get(int) */
+    private static final MethodHandle LIST_GET;
+    /** pointer to List.set(int,Object) */
+    private static final MethodHandle LIST_SET;
+    static {
+        Lookup lookup = MethodHandles.lookup();
```
Comment (uschindler) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/Def.java]: This should better be a MethodHandles.publicLookup(), as we only lookup public methods from public classes. This is not critical here, as its a static part only, but for "correctness".

**Comment by rmuir** [comment]
Also I want to defer trying to "improve" the signature we create here. Currently we often know the real types of parameters, e.g. `get(0)`, we know the one parameter is an int, we don't need to box it, but we explicitly do a lot of conversions anyway. That is because the current code is setup to work that way. It is silly to keep it at Object long-term but I want to fix the super slow calls first. 

We can later refactor to improve that (e.g. all tests pass if i do this: https://github.com/rmuir/elasticsearch/commit/e029b9ec546f872b7ee093a9ef8d6180b9709318 but thats really hacky).

**Comment by clintongormley** [comment]
This is phenomenal - nice work!

**Comment by s1monw** [comment]
> This is phenomenal - nice work!

++

**Comment by uschindler** [comment]
Ho Robert, I have seen you made the suggested changes already. Thanks! To me this looks fine - I only reviewed the CallSite part. Looks great.

I would also do the other optimizations in separate issues, like Array.getLength() (which may be slower than a real access to the length field in bytecode) to some static&private getters (see your TODO). The problem is that getting the array length is only a property in Java syntax, but behind the scenes all is a separate bytecode for array load/store/length: https://en.wikipedia.org/wiki/Java_bytecode_instruction_listings

**Comment by uschindler** [comment]
P.S.: Thanks for the fruitful discussions this weekend! The whole MethodHandle stuff is so great. I was happy to help and add suggestions. I will review the remaining code a bit later and provide some feedback if there are questions open.

**Comment by mikemccand** [comment]
We are now running simple script search performance tests in the nightly classic benchmarks: https://benchmarks.elastic.co/index.html#search_qps_scripts

It shows already a big speedup (I will add annots!) for painless dynamic from the last few improvements.

**Comment by uschindler** [comment]
Hi Robert,
I created the MethodHandle code for the array.length bytecode: https://gist.github.com/uschindler/916930741f245204e418c56025a8b58a

This code has a static method where you can pass an primitive or x-extends-Object array and you get a MethodHandle to retun its length. This should be faster than the reflective Array.getLength (because it is statically compiled and we have a separate method for every primitive type). Maybe we shold benchmark it.

Basically it holds a static map array->type to methodhandle (dynamically build) for primitives and Object[], and if the actual type is not found there (e.g, String[]), it adds a necessary cast to Object[].

The code has test as main() method inline.

**Comment by rmuir** [comment]
Thanks Uwe, I will fold in your ArrayUtil class instead of calling the slow Array.getLength.

**Comment by jdconrad** [comment]
LGTM.  Thanks for doing this!

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/WriterExternal.java [hunk 4]
```diff
@@ -729,31 +701,22 @@ private void writeCallExternal(final ExtcallContext source) {
...
-            execute.invokeStatic(definition.defobjType.type, DEF_METHOD_CALL);
+            writeDynamicCallExternal(source);
+        }
+    }
+    
+    private void writeDynamicCallExternal(final ExtcallContext source) {
+        final ExtNodeMetadata sourceenmd = metadata.getExtNodeMetadata(source);
+        final List<ExpressionContext> arguments = source.arguments().expression();
+        
+        StringBuilder signature = new StringBuilder();
```
Comment (uschindler) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/WriterExternal.java]: I would build the signature with the ASM Type class maybe... Bt this is also fine - I just hate those hardcoded stringified bytecode constants!

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/WriterConstants.java [hunk 5]
```diff
@@ -40,22 +44,25 @@
...
 
     final static Type SCORE_ACCESSOR_TYPE    = Type.getType(ScoreAccessor.class);
     final static Method SCORE_ACCESSOR_FLOAT = getAsmMethod(float.class, "floatValue");
 
-    final static Method DEF_METHOD_CALL = getAsmMethod(
-        Object.class, "methodCall", Object.class, String.class, Definition.class, Object[].class, boolean[].class);
+    /** dynamic callsite bootstrap signature */
+    final static MethodType DEF_BOOTSTRAP_TYPE = MethodType.methodType(CallSite.class, MethodHandles.Lookup.class, 
+                                                                       String.class, MethodType.class, int.class);
+    final static Handle DEF_BOOTSTRAP_HANDLE = new Handle(Opcodes.H_INVOKESTATIC, "org/elasticsearch/painless/DynamicCallSite",
```
Comment (uschindler) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/WriterConstants.java]: I'd use `Type.getInternalName(DynamicCallSite.class)` instead of constant. This allows to rename class if needed and not break static string constants

**Comment by rmuir** [comment]
@uschindler for the array length, i want to defer that to another issue. I don't want us having possibility of slowness, but Array.getLength is an intrinsic method, it may not really be slow. We can benchmark it to see.

**Comment by danielmitterdorfer** [comment]
@rmuir that's a great improvement. :) I think that besides looking at the median, it's also worth looking at the percentile distribution (esp. higher percentiles like 99%, 99.9%) to get a better overall impression. Just out of curiosity: Did you have a chance to have a look at these too?

**Comment by rmuir** [comment]
@danielmitterdorfer The nightly benchmark is just to explain what the differences mean at a high level, it is not really what I used to develop the changes. That benchmark is extremely noisy and only outputs the median.

This isn't a micro-optimization, I didn't tune anything with any benchmark. A branch in the code + exact methodhandle call is night and day vs looking up a methodhandle from our whitelist and adapting it, and the situation is well studied and understood. Its just important to use correct datastructures for the problem (e.g. polymorphic cache) so that we don't call lookups over and over.

**Comment by danielmitterdorfer** [comment]
@rmuir Oh, I didn't expect that you'd use the benchmark to tune anything (and I also know that they're quite noisy. I think this is partly system-inherent but we should (a) also ensure that we eliminate as much noise as possible on the platform where we run nightlies and (b) adopt approaches that help us to deal with noise (both topics are on my todo list)). Thanks for your explanation.

**Comment by rmuir** [comment]
The nightlies could use big improvement. I think traditionally effort has been on indexing performance and not query performance. All the queries are noisy and nothing is setup for e.g. a developer to iterate/explore with (means not reindexing every time you run the benchmark, and being reasonably fast). 

For scripts, we also only test a fictional search script to at least have something. But for example, we don't even have a fictional update script yet, forget about having any variety in the types of scripts we want to benchmark or anything like that.

**Comment by uschindler** [comment]
> @uschindler for the array length, i want to defer that to another issue. I don't want us having possibility of slowness, but Array.getLength is an intrinsic method, it may not really be slow. We can benchmark it to see.

Yes, there should be a separate issue abbout the arrays and array-like structures. The arrayLoad and arrayStore methods in Def can be removed completely and similar handled by invokeDynamic. This is especially important because the reflective Array.get/Array.set are slow like hell (see http://stackoverflow.com/questions/30306160/performance-of-java-lang-reflect-array, https://bugs.openjdk.java.net/browse/JDK-8051447).

If we use invokeDyanmic here, too the bootsrap method can select the right method handles correctly based on types. Partly it is already implemented for the map gets, this can be done in a similar way for the list get/sets and using MethodHandles.arrayElementGetter/Setter. This should bring a huge improvement for array accesses!

**Comment by uschindler** [comment]
+1 as a first step. LGTM

**Comment by rmuir** [comment]
Thanks @uschindler for reviewing.

**Comment by danielmitterdorfer** [comment]
@rmuir thanks for your feedback. I have created elastic/rally#96 and elastic/rally#97. Feel free to create more tickets if you find more shortcomings or just approach me. I really appreciate your input.

**Comment by rmuir** [comment]
@danielmitterdorfer thanks for opening those issues! FYI I hope it didn't come across as a criticism of rally, just our current state, lots of room for improvement.

**Comment by danielmitterdorfer** [comment]
@rmuir You're welcome. Your feedback did not come as criticism at all, quite the contrary. I know that we have a lot of room for improvement and I also have lots of ideas on how to improve the situation but I really value your feedback.
